cli/src/commands/git.rs

```diff
@@ -848,13 +785,78 @@ fn cmd_git_push(
+                    Ok(Some(update)) => branch_updates.push((branch_name.to_owned(), update)),
+                    Ok(None) => writeln!(
+                        ui.stderr(),
+                        "Branch {branch_name}@{remote} already matches {branch_name}",
+                    )?,
+                    Err(reason) => return Err(reason.into()),
+                }
+            }
+
+            for (change_str, commit) in std::iter::zip(args.change.iter(), change_commits) {
+                let mut branch_name = format!(
+                    "{}{}",
+                    command.settings().push_branch_prefix(),
+                    commit.change_id().hex()
+                );
+                if !seen_branches.insert(branch_name.clone()) {
+                    continue;
+                }
+                let view = tx.base_repo().view();
+                if view.get_local_branch(&branch_name).is_absent() {
+                    // A local branch with the full change ID doesn't exist already, so use the
+                    // short ID if it's not ambiguous (which it shouldn't be most of the time).
+                    let short_change_id = short_change_hash(commit.change_id());
+                    if tx
+                        .base_workspace_helper()
+                        .resolve_single_rev(&short_change_id, ui)
+                        .is_ok()
+                    {
+                        // Short change ID is not ambiguous, so update the branch name to use it.
+                        branch_name = format!(
+                            "{}{}",
+                            command.settings().push_branch_prefix(),
+                            short_change_id
+                        );
+                    };
+                }
+                if view.get_local_branch(&branch_name).is_absent() {
+                    writeln!(
+                        ui.stderr(),
+                        "Creating branch {} for revision {}",
+                        branch_name,
+                        change_str.deref()
+                    )?;
+                }
+                tx.mut_repo()
+                    .set_local_branch_target(&branch_name, RefTarget::normal(commit.id().clone()));
+                let targets = TrackingRefPair {
+                    local_target: tx.repo().view().get_local_branch(&branch_name),
+                    remote_ref: tx.repo().view().get_remote_branch(&branch_name, &remote),
+                };
+                match classify_branch_update(&branch_name, &remote, targets) {
+                    Ok(Some(update)) => branch_updates.push((branch_name.clone(), update)),
+                    Ok(None) => writeln!(
+                        ui.stderr(),
+                        "Branch {branch_name}@{remote} already matches {branch_name}",
+                    )?,
+                    Err(reason) => return Err(reason.into()),
+                }
+            }
```

Comment:
Perhaps, the code block above can be moved to separate `if !use_default_revset`? It's unrelated to the return value of this `if` block.

```rust
if !use_default_revset {
   // branch name based resolution
}

// revset-based resolution
let revision_commit_id = if use_default_revset {
    ...
} else {
    ...
};